backend/search/models.py

```diff
@@ -40,54 +51,162 @@ class Meta:
     def __str__(self):
         return '"{}…" for {}'.format(self.xpath[:10], self.component)
 
+    def _get_cache_path(self, create_dir: bool) -> pathlib.Path:
+        """Get the Path of the caching file corresponding to this
+        ComponentSearchResult. If create_dir is True, create the parent
+        directory if needed."""
+        try:
+            settings.CACHING_DIR.mkdir(exist_ok=True, parents=True)
+        except (FileExistsError, FileNotFoundError):
+            raise SearchError('Could not access or create caching directory')
+        return settings.CACHING_DIR / str(self.id)
+
+    def get_results(self) -> dict:
+        """Return results as a dict"""
+        cache_filename = str(self._get_cache_path(False))
+        try:
+            cache_file = open(cache_filename, 'r')
+        except FileNotFoundError:
+            logger.error(
+                'Could not open cache file for ComponentSearchResult {}: {}'
+                .format(self.id, cache_filename)
+            )
+            raise SearchError('Cache file not found')
+        results = cache_file.read()
+        cache_file.close()
+        self.last_accessed = timezone.now()
+        self.save()
+        return parse_search_result(results, self.component.slug)
+
+    def _truncate_results(self, results: str, number: int) -> str:
+        matches = list(re.finditer('<match>', results))
+        start_of_nth_match = matches[number].span()[0]
+        return results[:start_of_nth_match]
+
     def perform_search(self, query_id=None):
```

Comment:
there's a lot going on in this function, maybe try to break it down with comments or helper methods